### ISSUE TITLE
AI models for graphics editing

### ISSUE THREAD

**Issue by Keavon** [GraphiteEditor/Graphite#1694]
Below is a list of AI model disciplines that will be useful tools in the graphics editing process. Feel free to comment with ideas for more items missing from this list.

- Image generation (to turn inputs like text, images, masks, or other control data into a desired image)
    - Infill (to generate the content in missing areas based on a mask, or uncrop an image)
    - Style transfer (to adapt the subject of one image to the art style of another image)
- Depth estimation (to generate a depth map that can be used for many procedural effects)
- Segmentation (to automatically mask a subject or break a scene apart into multiple subjects)
- Relighting (to change the direction and color cast of the lighting on a scene or subject)
- Novel view synthesis (to alter the perspective angle or FoV of a subject)

**Comment by CheeksTheGeek** [comment]
Can I possibly submit a GSOC Proposal for this exact mega-task? 
It seems like an interesting domain to which I'd like to contribute, can see starting points for 4 out of 5 pointers (except Novel View Synthesis), and would very much like to learn in these fields, and seems like the timeline matches with that of GSOC? 

Though I can understand if the maintainers would like someone more specialized to be up for this task, if I get the opportunity, I'll definitely make it up to the quality standards maintained in this great project!

Let me know if it's possible and I'll start working on a proposal right away!

**Comment by Keavon** [comment]
@CheeksTheGeek please join the Discord and discuss this with me there. The task is described somewhat further in <https://graphite.art/volunteer/guide/student-projects/#machine-learning-architecture>. Just to set expectations, I will note that this requires a decently advanced level of understanding in deep learning concepts, including a fluent understanding of how to use libraries like PyTorch or similar to implement concepts based on your readings of research papers.

**Comment by akshitgaur2005** [comment]
Hello! I'm interested in contributing to this task and had a few questions:

1. Is this task still open for GSoC 2024 contributors, or has it already been assigned?
2. Is there a possibility of it being included in GSoC 2025?

I have experience with Python, PyTorch, and deep learning algorithms. I've also experimented with setting up and using Stable Diffusion, Segment Anything, and llama.cpp. While I don't have Rust experience yet, I'm committed to learning it as both the language and this project domain interest me greatly.
I plan to contribute regardless of GSoC status, but early confirmation would help me prioritize my Rust learning schedule.

**Comment by akshitgaur2005** [comment]
While waiting for your confirmation, I went ahead and researched a bit about what would be needed for implementing these tools.

1. **Upscaling** - Upscaling seems simple enough. Candle could be used to implement something like ESRGAN.
2. **Image Generation** - Candle provides a pre-written implementation of Stable Diffusion as an example which is quite performant in and of itself. Infilling and rest will take more effort to seemlessly integrate.
3. **Depth Estimation** - Models like MiDaS do exist but there does not seem to be any implementation in rust as of now. It should be possible to implement it in Candle and use the pre-trained model.
4. **Segmentation** - Implementation for Segment Anything already exists for candle.
5. **Relighting** - Models such as DiLightNet and IC-Light exist, but I will have to explore this avenue more before I can say anything with surety.
6. **Novel View Synthesis** - Found a few papers but have to look more, same for deconvolving a blur.

**Comment by Keavon** [comment]
Hi @akshitgaur2005. It might be best if we discuss this through Discord in the `#ai-ml` channel. I'll just say that this is an important topic, but also the most speculative because there are many possible approaches and no single straightforward path, which means it's more uncertain as to whether it works out in a way that ends up being useful to Graphite. A big part of it would be pathfinding the right approach and making me feel confident that it's a solid path forward that is likely to result in good long-term outcomes. Now is a good time to get started if you're interested in it because you'd be able to experiment with solutions and make a case for why they'd succeed as a GSoC project. Normally, the default state of this project within the scope of a GSoC proposal would be that it's a risky one and therefore might not top our list of priorities if we have other good proposals that have a more certain pathway towards success given the nature of limited applicant slots we can choose. That said, even though it's a hard problem, it's not one we can continue to neglect for years while the industry marches past us, so some R&D in this direction is of value regardless. I'll say, a few sample projects would be an initial implementation of Segment Anything 2 or Depth Anything, and ideally running in WGPU through either Candle or Burn. The latter seemed more aligned with our long-term goals, but is more removed from the Hugging Face community than Candle which may mean we'd miss out on the value of HF's research adjacency. Let's talk more on Discord if you'd like to dive into the weeds of things.

**Comment by Keavon** [comment]
Great example from [RGB↔X](https://zheng95z.github.io/publications/rgbx24?lid=7ibrcrhosh7a) of the useful workflows that are possible:

![rgbx24_famous](https://github.com/user-attachments/assets/a9054990-b99b-4fcc-bcfb-d72901522c14)

**Comment by j-jasz** [comment]
Segmentation model with Apache 2.0 license: https://github.com/xuebinqin/DIS

**Comment by AnonGitCat** [comment]
I feel like Image Generation would be too controversial to include. There needs to be at least a way to hide/disable it if you do go through with it

**Comment by clarence112** [comment]
Gonna be honest, the moment a tool adds generative AI, I stop using it. Graphite would be no exception.  
Why add something artists generally consider an insult to their skills to an artistic tool?

**Comment by gracen-writes-code** [comment]
I'd caution to not add any generative AI, or, if you're dead-set on going in this direction, make it some sort of extension. From what I've seen, most artists (and aspiring artists such as myself) consider generative AI to be morally questionable at best and outright appropriation of labor at worst. I'm all for using machine learning tools to fill in parts of the design/illustration workflow which are banal and require no thought or creativity, but keep in mind that such tools should be meant to support artists, not replace them.

If you ignored everything I've just written and you're only interested in creating a tool that corporations can use to lay off their graphic design teams or open source startups can use to generate vaguely appealing branding, please consider switching away from the `graphite.art` domain, as asking a neural network to think for you is not anything remotely close to art. This project doesn't seem to be headed in this direction, at least from what I've seen, so I hope this entire paragraph is proven unnecessary. But you can't fault me for being cynical.
